### PR TITLE
[tls] add missing built in cipher stat names

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Commit Message: Adds 3 missing (and removes one duplicate) stat name for TLS v1.3 cipher suites. 
Risk Level: Low
Testing: Added unit tests to test all suits in https://tools.ietf.org/html/rfc8446#appendix-B.4

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 1]
```diff
@@ -472,8 +472,9 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
 
   // Add hardcoded cipher suites from the TLS 1.3 spec:
   // https://tools.ietf.org/html/rfc8446
-  stat_name_set_->rememberBuiltins(
-      {"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_AES_128_GCM_SHA256"});
+  stat_name_set_->rememberBuiltins({"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384",
```
Comment (jmarantz) [inline source/extensions/transport_sockets/tls/context_impl.cc:475]: can you add the appendix anchor to the link above?

**Review by jmarantz — COMMENTED** [review]
thanks for doing this!

**Review by jmarantz — DISMISSED** [review]

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 2]
```diff
@@ -471,9 +471,10 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
   }
 
   // Add hardcoded cipher suites from the TLS 1.3 spec:
-  // https://tools.ietf.org/html/rfc8446
-  stat_name_set_->rememberBuiltins(
-      {"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_AES_128_GCM_SHA256"});
+  // https://tools.ietf.org/html/rfc8446#appendix-B.4
+  stat_name_set_->rememberBuiltins({"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384",
+                                    "TLS_AES_128_CCM_SHA256", "TLS_AES_256_CCM_8_SHA256",
```
Comment (PiotrSikora) [inline source/extensions/transport_sockets/tls/context_impl.cc:476]: `TLS_AES_128_CCM_SHA256` and `TLS_AES_256_CCM_8_SHA256` are not supported by BoringSSL, so we can drop them, IMHO.

Comment (jmarantz) [reply source/extensions/transport_sockets/tls/context_impl.cc:476]: That makes sense; if you remove those, can you just leave a comment indicating they are in-spec but not in BoringSSL.

Comment (asraa) [reply source/extensions/transport_sockets/tls/context_impl.cc:476]: Yep, done!

File: test/extensions/transport_sockets/tls/context_impl_test.cc [hunk 3]
```diff
@@ -1977,11 +1977,15 @@ class SslContextStatsTest : public SslContextImplTest {
...
   // Incrementing a value for a cipher that is part of the configuration works, and
   // we'll be able to find the value in the stats store.
-  context_->incCounter("ssl.ciphers", "ECDHE-ECDSA-AES256-GCM-SHA384");
-  Stats::CounterOptConstRef cipher =
-      store_.findCounterByString("ssl.ciphers.ECDHE-ECDSA-AES256-GCM-SHA384");
-  ASSERT_TRUE(cipher.has_value());
-  EXPECT_EQ(1, cipher->get().value());
+  for (const auto& cipher :
+       {"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256"}) {
+    // Test all built-in TLS v1.3 cipher suites https://tools.ietf.org/html/rfc8446#appendix-B.4.
```
Comment (PiotrSikora) [inline test/extensions/transport_sockets/tls/context_impl_test.cc:1982]: Nit: `s/all/supported/`

**Review by jmarantz — DISMISSED** [review]

**Review by PiotrSikora — APPROVED** [review]
LGTM sans nit

**Review by ggreenway — APPROVED** [review]

**Comment by asraa** [comment]
Sorry for the trouble, had a variable name issue in the test that broke CI.

**Review by jmarantz — APPROVED** [review]

**Review by PiotrSikora — APPROVED** [review]
Thanks!

File: test/extensions/transport_sockets/tls/context_impl_test.cc [hunk 4]
```diff
@@ -1995,9 +2000,9 @@ TEST_F(SslContextStatsTest, IncOnlyKnownCounters) {
   // fallback registration does not occur. So we test for the fallback only in
   // release builds.
 #ifdef NDEBUG
-  cipher = store_.findCounterByString("ssl.ciphers.fallback");
-  ASSERT_TRUE(cipher.has_value());
-  EXPECT_EQ(1, cipher->get().value());
+  stat = store_.findCounterByString("ssl.ciphers.fallback");
```
Comment (PiotrSikora) [inline test/extensions/transport_sockets/tls/context_impl_test.cc:2003]: `stat` is not defined in this scope, you need to add `Stats::CounterOptConstRef ...`.

**Comment by PiotrSikora** [comment]
I think we need to spin 1.17.1 once this is merged, due to use of `ENVOY_BUG` in #14534.

/backport

**Comment by PiotrSikora** [comment]
> I think we need to spin 1.17.1 once this is merged, due to use of `ENVOY_BUG` in #14534.
> 
> /backport

cc @jmarantz @mattklein123

**Comment by jmarantz** [comment]
I think that's probably right. Release builds will work (will log the cipher-name warning but with decreasing frequency I think). But we support debug docker images and they will crash.

File: test/extensions/transport_sockets/tls/context_impl_test.cc (hunk 4 above)
Comment (jmarantz) [reply test/extensions/transport_sockets/tls/context_impl_test.cc:2003]: yeah when hand-testing this you want to run this unit test with "-c opt"

**Comment by mattklein123** [comment]
Check CI and merge main?

/wait

File: test/extensions/transport_sockets/tls/context_impl_test.cc (hunk 4 above)
Comment (asraa) [reply test/extensions/transport_sockets/tls/context_impl_test.cc:2003]: Yeah :/ I had only tested in debug and fastbuild.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**